test/TurnUITests.java

```diff
@@ -2,29 +2,99 @@
+        TurnUI turns = new TurnUI(printer, System.out, in);
+
+        turns.takeTurn(board, game);
+        turns.takeTurn(board, game);
+
+        assertEquals(board.markerAt(1), "O");
+    }
+
+    @Test
+    public void asksForInputAgainIfInputIsText() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("invalidInput\n2").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
+
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
+    }
+
+    @Test
+    public void asksForInputAgainIfInputIsTooLarge() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("9999999999999999999999999999\n7").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
 
-        turns.takeTurn(board, printer, game, in);
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
+    }
 
-        in = new ByteArrayInputStream("2".getBytes());
-        turns.takeTurn(board, printer, game, in);
+    @Test
+    public void asksForInputAgainIfInputIsNotOnBoard() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("10\n7").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        TurnUI turns = new TurnUI(printer, out, in);
 
-        assertEquals(board.markerAt(2), "O");
+        turns.takeTurn(board, game);
 
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
     }
 
+    @Test
+    public void asksForInputAgainIfSelectionTaken() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("5\n2").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        board.placeMarker(4, "X");
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
 
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
+    }
+
+    @Test
+    public void zeroIndexSelection() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("5").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
+
+        assertEquals(board.isOccupied(4), true);
+
```

Comment:
Whitespace.

Lots of nitpicks in this PR. In general, looks pretty good. Testing validation logic via the UI is the big thing that stands out to me.